src/DotVVM.Framework/Compilation/Binding/MemberExpressionFactory.cs

```diff
@@ -408,17 +408,53 @@ private Type GetGenericParameterType(Type genericArg, Type[] searchedGenericType
                 }
                 else if (sgt.IsGenericType)
                 {
-                    Type[] genericArguments;
+                    Type[] genericArguments = null;
                     var expression = expressionTypes[i];
 
-                    // Arrays need to be handled in a special way to obtain instantiation
                     if (expression.IsArray)
+                    {
+                        // Arrays need to be handled in a special way to obtain instantiation
                         genericArguments = new[] { expression.GetElementType() };
+                    }
                     else
-                        genericArguments = expression.GetGenericArguments();
+                    {
+                        if (expression.IsGenericType && sgt.GetGenericTypeDefinition() == expression.GetGenericTypeDefinition())
+                        {
+                            // We have exactly the same type => return generic arguments
+                            genericArguments = expression.GetGenericArguments();
+                        }
+                        else if (sgt.IsInterface)
+                        {
+                            // We must find the instantiation within an implemented generic interface
+                            var implementation = expression.GetInterfaces().Where(ifc => ifc.IsGenericType && ifc.GetGenericTypeDefinition() == sgt.GetGenericTypeDefinition()).Take(2).ToList();
```

Comment:
Is it actually possible for `expression` to implement more than one interface that is equal to the generic type definition of `sgt`? If such a thing ever happened, wouldn't it mean that something went horribly wrong and would warrant an exception?